### ISSUE TITLE
Support new MLJ iteration interface

### ISSUE THREAD

**Issue by ablaom** [Evovest/EvoTrees.jl#74]
Training losses don't seem to be exposed in the MLJ interface. This would be nice. I'm close to finishing up an iterative control wrapper for MLJ models and if they report training losses, one can apply "training progress" modified stopping criterion, such as [PQ](https://github.com/ablaom/EarlyStopping.jl#criteria)

**Comment by ablaom** [comment]
@jeremiedb I would love to use EvoTrees to test and demonstrate the new iterative model wrapper for MLJ. For full integration this would require EvoTrees to implement the following traits in `MMI=MLJModelInterface`:

```julia
MMI.supports_training_losses(model::EvoTreeModel) = true
MMI.training_losses(model::EvoTreeModel, report) = vector of training losses
MMI.iteration_parameter(model::EvoTreeModel) = :nrounds
```

I can make the necessary PR if you put the training losses into the MLJ report object, say as `report=(feature_importances=..., training_losses=...)`

The length of `training_losses` should be `nrounds`. Your `update` method should update the training_losses appropriately , so you might need to pass them to `cache`.

**Comment by jeremiedb** [comment]
@ablaom There are some aspects of the design I'd like to clarify. In particular, how the tracking of the metric on evaluation will be performed. The reason for this is that my go-to usage of metric tracking is only performed on the evaluation data, I even often don't even track the training data. 
As such, when it comes to tracking metric, I would tend to view it as a something that is ideally kept in a seperate compartment to the model training functions. For example:
```
fit(model, cache)
update_metric!(train_metric, model, pred_train, y_eval)
update_metric!(eval_metric, model, pred_eval, y_train)
```

Naturally, the pred_train/eval and y_train/eval could all be contained in the model cache. Actually I think this would be the best way to go (and could even be extended to a collection of eval datas). However, I'm unclear whether the MLJ could be set so as to handle eval data. Could something like: `machine(tree_model, X, y, X_eval = X_eval, y_eval=y_eval)` be implemented ? 
For algos such as EvoTrees/XGB & cie, ability to manage the early stopping on an evaluation dataset is critical, so I think figuring out first how such eval metric would be handled is desirable.

**Comment by ablaom** [comment]
Thanks for looking at this. Perhaps your questions will be answered by looking at the doc-string for my WIP IteratedModel wrapper, copied below. See also an EvoTrees example  [here](https://github.com/ablaom/MLJIteration/blob/master/examples/tree_booster/tree_booster.jl).

I totally agree that training losses are not a subsitute for an out-of-sample loss, and that's the whole point of my wrapper, which allows you to specify a train/test split using the `resampling` hyperparameter.  However, in the PQ stopping criterion of [Prechel 1998](https://link.springer.com/chapter/10.1007%2F3-540-49430-8_3) one additionally takes training losses into account to modify the generalization error estimate according to recent "training progress" (no point waiting for the generalization loss to go up a lot if training error has plateaued). In the example linked above, I don't use this criterion, because you haven't made these available. Also, I have to manually declare the name of the iteration parameter. 

> Could something like: machine(tree_model, X, y, X_eval = X_eval, y_eval=y_eval) be implemented ?

No, the idea is that the splitting happens under the hood and is controlled by the `resampling` hyper-parameter of the wrapper - just like in MLJ tuning. This approach is more compositional. 

Is this clearer?
---

    IteratedModel(model=nothing,
                  controls=[Step(10), Patience(5), NumberLimit(50), NotANumber()],
                  final_train=false,
                  resampling=Holdout(),
                  measure=nothing,
                  weights=nothing,
                  class_weights=nothing,
                  class_weights=nothing,
                  operation=predict,
                  verbosity=1,
                  check_measure=true,
                  iteration_parameter=nothing,
                  cache=true)

Wrap the specified `model <: Supervised` in the specified iteration
`controls`. Training a machine bound to the wrapper iterates a
corresonding machine bound to `model`. Here `model` should support
iteration. Controls are discussed further below.

To make out-of-sample losses available to the controls, the machine
bound to `model` is only trained on part of the data, as iteration
proceeds.  See details on training below. Specify `final_train=true`
to ensure the model is retrained on *all* available data, using the
same number of iterations, once controlled iteration has stopped.

Specify `resampling=nothing` if all data is to be used for controlled
iteration, with each out-of-sample loss replaced by the most recent
training loss, assuming this is made available by the model
(`supports_training_losses(model) == true`). Otherwise, `resampling`
must have type `Holdout`.

Assuming `final_train=true` or `resampling=nothing`,
`iterated_model` behaves exactly like the original `model` but with
the iteration parameter automatically selected. If
`final_train=false` (default) and `resampling` is not `nothing`, then
`iterated_model` behaves like the original model trained on a subset
of the provided data.

Controlled iteration can be continued with new `fit!` calls (warm
restart) by mutating a control, or by mutating the iteration parameter
of `model`, which is otherwise ignored.


### Training

Given an instance `iterated_model` of `IteratedModel`, calling
`fit!(mach)` on a machine `mach = machine(iterated_model, data...)`
performs the following actions:

- Assuming `resampling !== nothing`, the `data` is split into *train* and
  *test* sets, according to the specified `resampling` strategy, which
  must have type `Holdout`.

- A clone of the wrapped model, `iterated_model.model`, is bound to
  the train data in an internal machine, `train_mach`. If `resampling
  === nothing`, all data is used instead. This machine is the object
  to which controls are applied. For example, `Callback(fitted_params
  |> print)` will print the value of `fitted_params(train_mach)`.

- The iteration parameter of the clone is set to `0`.

- The specified `controls` are repeatedly applied to `train_mach` in
  sequence, until one of the controls triggers a stop. Loss-based
  controls (eg, `Patience()`, `GL()`, `Threshold(0.001)`) use an
  out-of-sample loss, obtained by applying `measure` to predictions
  and the test target values. (Specifically, these predictions are
  those returned by `operation(train_mach)`.)  If `resampling ===
  nothing` then the most recent training loss is used instead. Some
  controls require *both* out-of-sample and training losses (eg,
  `PQ()`).

- Once a stop has been triggered, a clone of `model` is bound to all
  `data` in a machine called `mach_production` below, unless
  `final_train == false` or `resampling === nothing`, in which case
  `mach_production` coincides with `train_mach`.


### Prediction

Calling `predict(mach, Xnew)` returns `predict(mach_production,
Xnew)`. Similar similar statements hold for `predict_mean`,
`predict_mode`, `predict_median`.


### Controls

Every set of controls should include a `Step(...)`, usually the first control.

For a summary of all available controls, see
[https://github.com/ablaom/IterationControl.jl#controls-provided](https://github.com/ablaom/IterationControl.jl#controls-provided). Basic
controls for getting started are summarized below. Query the
doc-strings for details and advanced options.

control                          | description
---------------------------------|----------------------------------
`Step(n=1)`                      | Train model for `n` iterations
`Info(f=identity)`               | Log to `Info` the value of `f(train_mach)`
`Callback(f=_->nothing)`         | Call `f(train_mach)`
`TimeLimit(t=0.5)`               | Stop after `t` hours
`NumberLimit(n=100)`             | Stop after `n` control cycles
`WithLossDo(f=x->@info(x))`      | Call `f(loss)` where `loss` is current loss
`NotANumber()`                   | Stop when `NaN` encountered
`Threshold(value=0.0)`           | Stop when `loss < value`
`Patience(n=5)`                  | Stop after `n` consecutive loss increases
`Save(filename="machine.jlso")`  | Save `train_mach` machine to file
`MLJIteration.skip(control, p=1)`| Apply `control` but only every `p` cycles.

A control is permitted to mutate the fields (hyper-parameters) of
`train_mach.model` (the clone of `model`). For example, to mutate a
learning rate one might use the control

    Callback(m -> m.model.eta = 1.05*m.model.eta)

However, unless `model` supports warm restarts with respect to changes
in that parameter, this will trigger retraining of `train_mach` from
scratch, with a different training outcome, which is not recommended.

See the MLJ documentation for user-defined controls.


### Warm restarts

If `iterated_model` is mutated and `fit!(mach)` is called again, then
a warm restart is attempted if the only parameters to change are
`model` or `controls` or both.

Specifically, `train_mach.model` is mutated to match the current value
of `iterated_model.model` and the iteration parameter of the latter is
updated to the last value used in the preceding `fit!(mach)` call. Then
repeated application of the (updated) controls begin anew.

**Comment by ablaom** [comment]
And, yes, you can control how often an out-of-sample loss is calculated, through the  size of `n` in the control `Step(n)`.

**Comment by jeremiedb** [comment]
Thanks for these clarifications. 

Looking at the booster example; 
```julia
imodel = IteratedModel(model=model,
                       resampling=Holdout(rng=123),
                       measures=[brier_loss, log_loss],
                       iteration_parameter=:nrounds,
                       controls=[Step(2),
                                 WithIterationsDo(i->push!(iterations, i)),
                                 WithLossDo(update_plot),
                                 GL(200),
                                 TimeLimit(Second(30))])
```

I'm still struggling to understand why the model report should carry the fitting information. 

What I understand from the IteratedModel logic, is that when `fit!` is called, the logic is essentially: 

 1. X,y data is split into x_train, x_eval, y_train, y_eval according to `resampling`. 
 2. model is fit on x_train,y_train for 1 iteration
 3. prediction is made on x_eval => pred_eval / y_eval
 4. measures are calculated on x_eval
 5. decision to continue training is made according to the `controls`

When it comes to the the measures on the training data, why not opt add the same steps 3 & 4 but for the training data? As I understand it, if the measures are to be calculated within the EvoTrees fit function, then it seems to involve that there would be the need to EvoTrees to support MLJ available measures such as `brier` or `hinge`. I would much prefer avoiding too much constraint on the format and features of the library, and handle the wrapping through the essential functions such as `fit` and `predict` as it has been so far.

What I understand from your proposal is: 

```
 1. X,y data is split into x_train, x_eval, y_train, y_eval according to `resampling`. 
 2. model is fit on x_train,y_train for 1 iteration
 3. prediction is made on x_eval => pred_eval
 4. measures for eval are calculated on from pred_eval / y_eval
 5. measures for train data are fetched from the the model report
 6. decision to continue training is made according to the `controls`
```

While what seems less "invasive" from my perspective is : 

```
 1. X,y data is split into x_train, x_eval, y_train, y_eval according to `resampling`. 
 2. model is fit on x_train,y_train for 1 iteration
 3. prediction is made on x_eval => pred_eval
 4. predictions for train are fetched from model's cache => pred_train
 5. measures for eval are calculated on pred_eval / y_eval
 6. measures for train are calculated on pred_train / y_train
 7. decision to continue training is made according to the `controls`
```

Hope the above may help clarify where there might be some misunderstanding. 

Also, maybe I overlooked the docs describing it, but how does the iterative process is expected to handle cases where there are more than 1 measure? Ie. does training stop whenever criteria for one measure is met, or when criteria for all measures are met? 

Rapidly, I'd recommend having a criteria `iter since best`, which is the vanilla one uses by EvoTrees but also XGboost and LightGBM, which seems to be missing. 

From a personal perspective, I understand the need for having utility functions that wrap common operation such as this iterative training and over multiple CVs, though in the meantime, the need for a very intuitive /  transparent behavior of what is happening will be critical in my opinion for adoption. Otherwise, I think many may resort to some custom loops iterating over CVs and hyper params. I guess that resonates in some way with recent debates in Flux around the need for the current lightweight train! helper function vs none vs a more fully fledge one. 
From my experience, if the operations can me broken down into simple/intuitive building blocks, it may have better chance to suits the various and unpredictable needs of users who will compose them vs. attempting to have single entry point that tries to take care of all scenarios.

**Comment by ablaom** [comment]
Thanks for taking the time to look into some of these details. I think I can see where our wires are crossed 😄 Thanks for your patience. 

> When it comes to the the measures on the training data, why not opt add the same steps 3 & 4 but for the training data? A

I should have said that one should only include training losses in the report *if these are being computed anyway, as a biproduct of training*. If not, I doubt the extra expense of computing them post-facto (in the natural way you suggest here) is worthwhile, just so you can use PQ. 

If EvoTrees generates these training losses for free, it would nice to make them available. If not, then that's no big deal, one can live without the PQ option. 

> Also, maybe I overlooked the docs describing it, but how does the iterative process is expected to handle cases where there are more than 1 measure? Ie. does training stop whenever criteria for one measure is met, or when criteria for all measures are met?

Good question. This can be handled with the `WithEvaluationDo(f=..., stop_if_true=true)` control (not mentioned in the doc-string). This control stops when the predicate `f`, applied to the current evaluation object, is `true`. The "evaluation object" is essentially the output of `MLJBase.evaluate!(model, Xtest, ytest, ...)` with all the options (eg, `measures=...`) specified as hyperparameters of the `IteratedModel` wrapper. Recall that this object encodes all the measurements for each measure and other stuff. So, as long as you are familiar with how to access this object, you can design `f`  accordingly. For example, if you have specified two measures with `measures=[fscore, brier_score]`, a facetious choice might be:

```julia
function f(e)
    e.measurement[1] > 0.01 || e.measurement[2] > 0.5
end
````

But a more useful criterion might require the user to specify a custom control, which is possible and will be documented.

> Rapidly, I'd recommend having a criteria iter since best, which is the vanilla one uses by EvoTrees but also XGboost and LightGBM, which seems to be missing.

Sure. So stop when the number of iterations since  the best so-far exceeds `iter`?

**Comment by jeremiedb** [comment]
> I doubt the extra expense of computing them post-facto

As you see it fits! Currently, the `fit` function for EvoTrees doesn't consider any metrics at all. However, just like XGBoost if I'm not mistaken, it does't carry the training predictions, so I think it would be a more reasonable assumption for the support PQ to expect the cache to contain the training predictions than the report to track the metrics specified in measures (which aren't passed to the `fit` function anyway if I'm not mistaken, making its application even harder). 

In this context, what I would add to the MLJ wrapper would then be the follwoing: 
```julia
MLJModelInterface.supports_training_losses(::EvoTypes) = false # which maybe could be substitute for "supports_training_pred" ?
MLJModelInterface.iteration_parameter(::EvoTypes) = :nrounds
```

> Sure. So stop when the number of iterations since the best so-far exceeds `iter`?

Exactly! 

On a sidenote, I've added the features_importance to the report, however, of the `reformat`, the features names gets lost since the fit function received a Matrix. Hence, features names are vanilla generated with: 
`report = (feature_importances = importance(fitresult, ["feat_$i" for i in 1:size(X, 2)]), )` which results in a certainly less desirable outcome. Would you see a way to pass through the features names despite the `reformat`?

**Comment by ablaom** [comment]
> As you see it fits! Currently, the fit function for EvoTrees doesn't consider any metrics at all. However, just like XGBoost if I'm not mistaken, it does carry the training predictions, so I think it would be a more reasonable assumption for the support of PQ to expect the cache to contain the training predictions than the report to track the metrics specified in measures (which aren't passed to the fit function anyway if I'm not mistaken, making its application even harder).

As I thought training losses might be something the user is interested in, it should go in the report. In the cache it is not user-inspectable. I don't think we need to make training predictions user accessible, but nor are they enough by themselves for the IteratedModel wrapper to make use of as presently designed. 

So I propose EvoTrees add only
> ```julia
> MLJModelInterface.iteration_parameter(::EvoTypes) = :nrounds
> ```
The other line you mention is the fallback (`false`) so not needed. 


> 
> On a sidenote, I've added the features_importance to the report, however, of the `reformat`, the features names gets lost since the fit function received a Matrix. Hence, features names are vanilla generated with:
> `report = (feature_importances = importance(fitresult, ["feat_$i" for i in 1:size(X, 2)]), )` which results in a certainly less desirable outcome. Would you see a way to pass through the features names despite the `reformat`?

Yes, I saw this. To improve this I suggest you make the feature names part of your model-specific representation of the data. Something like:

```julia
MLJModelInterface.reformat(::EvoTypes, X, y) = ((matrix=MLJModelInterface.matrix(X), names=schema(X).names), y)
MLJModelInterface.reformat(::EvoTypes, X) = ((matrix=MLJModelInterface.matrix(X), names=schema(X).names), )

MLJModelInterface.selectrows(::EvoTypes, I, A, y) = ((matrix=view(A.matrix, I, :), names=A.names), view(y, I))
MLJModelInterface.selectrows(::EvoTypes, I, A) = ((matrix=view(A.matrix, I, :), names=A.names),)
```
and adjust `fit` and `predict` accordingly. Then `fit` has access to the names (`A.names`) and you can splice them into your feature importances. 

This is a common scenario - I should probably adjust the example in the docs to reflect this.

**Comment by jeremiedb** [comment]
I got mixed results with the above approach for handling the feature names. 
In particular, when it comes to: 
```julia
julia> pred_train = predict(tree, selectrows(tree_model, train, X))
ERROR: type NamedTuple has no field matrix
```

Not exactly sure why even attempting to fix with `reformat` still doesn't work: 
```julia
julia> pred_train = predict(tree, selectrows(tree_model, train, reformat(tree_model, X)))
ERROR: type Tuple has no field matrix
```

Anyhow, I have some reserves on what seems to me like unexpected side effects: I think it's reasonable to expect that a call to select_rows do only returns a subset of rows, and not assumes some handling of associated feature names. 

I've lloked at alternatives such as dispatching the on `fit` according to input X so that features names could be handled in the internal EvoTrees routine, but without success as the reformat/select rows are all called upstream: 

```julia
function MLJModelInterface.fit(model::EvoTypes, verbosity::Int, X, y)
    names = [name for name ∈ schema(X).names]
    fitresult, cache = init_evotree(model, MLJModelInterface.matrix(X), y, verbosity=verbosity)
    grow_evotree!(fitresult, cache, verbosity=verbosity)
    report = (feature_importances = importance(fitresult, names), )
    return fitresult, cache, report
end

function MLJModelInterface.fit(model::EvoTypes, verbosity::Int, X::AbstractMatrix, y)
    names = ["feat_$i" for i in 1:size(X, 2)]
    fitresult, cache = init_evotree(model, X, y, verbosity=verbosity)
    grow_evotree!(fitresult, cache, verbosity=verbosity)
    report = (feature_importances = importance(fitresult, names), )
    return fitresult, cache, report
end
```

In short, I remain a little doubtful about behavior such as `select_row`which carry on features names as my impression is does kind of behavior add to the mental charge in contrast to functions with minimal side effects. 
The way I would go within EvoTrees own API if I was to add support for Tables/DataFrames input would be to first extract the name based on the nature of the input (generic names for Matrix or real ones for DataFrames), and then continue on the common routine that assumes a Matrix input. It would be fairly lightweight on my end to support both DataFrames and Matrix, so it could be a win-win.

**Comment by jeremiedb** [comment]
@ablaom I've played around the options for the integration features names on different input types, and I couldn't find a better way to go than what you proposed :D 
So I just pushed a new version of EvoTrees, v0.7.0, which include the latest discussion points. As well, GPU support for training is now directly supported through the `device = gpu` option in the model parameters.

**Comment by ablaom** [comment]
Yes, and thanks for persevering with the interface.
